File: examples/go/error_handling.go
// Result-style error handling and fail-closed patterns.
package examples

import (
	"errors"
	"fmt"
)

// --- Custom error types ---

type AuthError struct {
	Reason string
	UserID int
}

func (e *AuthError) Error() string {
	return fmt.Sprintf("auth denied for user %d: %s", e.UserID, e.Reason)
}

type NotFoundError struct {
	Resource string
	ID       int
}

func (e *NotFoundError) Error() string {
	return fmt.Sprintf("%s %d not found", e.Resource, e.ID)
}

// --- Core logic with structured errors ---

type UserID struct{ Value int }

type Permission struct {
	UserID UserID
	Roles  []string
}

func ParseUserID(raw int) (UserID, error) {
	if raw < 1 {
		return UserID{}, fmt.Errorf("user id must be positive, got %d", raw)
	}
	return UserID{Value: raw}, nil
}

var knownPerms = map[int][]string{
	1: {"read", "write"},
	2: {"read"},
}

func LookupPermissions(id UserID) (Permission, error) {
	roles, ok := knownPerms[id.Value]
	if !ok {
		return Permission{}, &NotFoundError{Resource: "user", ID: id.Value}
	}
	return Permission{UserID: id, Roles: roles}, nil
}

func hasRole(roles []string, required string) bool {
	for _, r := range roles {
		if r == required {
			return true
		}
	}
	return false
}

// Authorize demonstrates fail-closed: deny on any error.
func Authorize(rawID int, required string) error {
	id, err := ParseUserID(rawID)
	if err != nil {
		return &AuthError{Reason: err.Error(), UserID: rawID}
	}

	perm, err := LookupPermissions(id)
	if err != nil {
		var nf *NotFoundError
		if errors.As(err, &nf) {
			return &AuthError{Reason: "unknown user", UserID: rawID}
		}
		return &AuthError{Reason: "lookup failed", UserID: rawID}
	}

	if !hasRole(perm.Roles, required) {
		return &AuthError{Reason: fmt.Sprintf("missing %q role", required), UserID: rawID}
	}
	return nil
}

File: examples/rust/error_handling.rs
//! Result types and fail-closed error handling.
#![allow(dead_code)]

use std::fmt;

// --- Custom error types ---

#[derive(Debug)]
enum AppError {
    Parse { field: String, message: String },
    NotFound { resource: String, id: u64 },
    AccessDenied { reason: String },
}

impl fmt::Display for AppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Parse { field, message } => write!(f, "parse error in {field}: {message}"),
            Self::NotFound { resource, id } => write!(f, "{resource} {id} not found"),
            Self::AccessDenied { reason } => write!(f, "access denied: {reason}"),
        }
    }
}

impl std::error::Error for AppError {}

// --- Domain types ---

#[derive(Debug, Clone)]
struct UserId(u64);

// --- Boundary: parse ---

fn parse_user_id(raw: &str) -> Result<UserId, AppError> {
    let n: u64 = raw.trim().parse().map_err(|_| AppError::Parse {
        field: "user_id".into(),
        message: format!("not a valid number: {raw:?}"),
    })?;
    if n == 0 {
        return Err(AppError::Parse {
            field: "user_id".into(),
            message: "must be positive".into(),
        });
    }
    Ok(UserId(n))
}

// --- Core: lookup ---

fn lookup_roles(id: &UserId) -> Result<Vec<String>, AppError> {
    match id.0 {
        1 => Ok(vec!["read".into(), "write".into()]),
        2 => Ok(vec!["read".into()]),
        _ => Err(AppError::NotFound {
            resource: "user".into(),
            id: id.0,
        }),
    }
}

// --- Fail-closed: deny by default on any error ---

fn authorize(raw_id: &str, required: &str) -> Result<(), AppError> {
    let id = parse_user_id(raw_id)?;
    let roles = lookup_roles(&id)?;

    if !roles.iter().any(|r| r == required) {
        return Err(AppError::AccessDenied {
            reason: format!("missing {:?} role", required),
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_authorize_success() {
        assert!(authorize("1", "read").is_ok());
        assert!(authorize("1", "write").is_ok());
    }

    #[test]
    fn test_authorize_denied() {
        assert!(authorize("2", "write").is_err());
        assert!(authorize("999", "read").is_err());
        assert!(authorize("abc", "read").is_err());
    }
}
